burn-book/src/building-blocks/dataset.md

```diff
@@ -133,3 +138,263 @@ multiple times over the dataset and only checkpoint when done. You can consider
+The process is illustrated in the figure below for the MNIST dataset.
+
+<img title="Burn Data Loading Pipeline" alt="Burn Data Loading Pipeline" src="./dataset.png">
+
+Although we have conveniently implemented the
+[`MnistDataset`](https://github.com/tracel-ai/burn/blob/main/crates/burn-dataset/src/vision/mnist.rs)
+used in the guide, we'll go over its implementation to demonstrate how the `Dataset` and `Batcher`
+traits are used.
+
+The [MNIST dataset](http://yann.lecun.com/exdb/mnist/) of handwritten digits has a training set of
+60,000 examples and a test set of 10,000 examples. A single item in the dataset is represented by a
+\\(28 \times 28\\) pixels black-and-white image (stored as raw bytes) with its corresponding label
+(a digit between \\(0\\) and \\(9\\)). This is defined by the `MnistItemRaw` struct.
+
+```rust, ignore
+# #[derive(Deserialize, Debug, Clone)]
+struct MnistItemRaw {
+    pub image_bytes: Vec<u8>,
+    pub label: u8,
+}
+```
+
+With single-channel images of such low resolution, the entire training and test sets can be loaded
+in memory at once. Therefore, we leverage the already existing `InMemDataset` to retrieve the raw
+images and labels data. At this point, the image data is still just a bunch of bytes, but we want to
+retrieve the _structured_ image data in its intended form. For that, we can define a `MapperDataset`
+that transforms the raw image bytes to a 2D array image (which we convert to float while we're at
+it).
+
+```rust, ignore
+const WIDTH: usize = 28;
+const HEIGHT: usize = 28;
+
+# /// MNIST item.
+# #[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct MnistItem {
+    /// Image as a 2D array of floats.
+    pub image: [[f32; WIDTH]; HEIGHT],
+
+    /// Label of the image.
+    pub label: u8,
+}
+
+struct BytesToImage;
+
+impl Mapper<MnistItemRaw, MnistItem> for BytesToImage {
+    /// Convert a raw MNIST item (image bytes) to a MNIST item (2D array image).
+    fn map(&self, item: &MnistItemRaw) -> MnistItem {
+        // Ensure the image dimensions are correct.
+        debug_assert_eq!(item.image_bytes.len(), WIDTH * HEIGHT);
+
+        // Convert the image to a 2D array of floats.
+        let mut image_array = [[0f32; WIDTH]; HEIGHT];
+        for (i, pixel) in item.image_bytes.iter().enumerate() {
+            let x = i % WIDTH;
+            let y = i / HEIGHT;
+            image_array[y][x] = *pixel as f32;
+        }
+
+        MnistItem {
+            image: image_array,
+            label: item.label,
+        }
+    }
+}
+
+type MappedDataset = MapperDataset<InMemDataset<MnistItemRaw>, BytesToImage, MnistItemRaw>;
+
+# /// The MNIST dataset consists of 70,000 28x28 black-and-white images in 10 classes (one for each digits), with 7,000
+# /// images per class. There are 60,000 training images and 10,000 test images.
+# ///
+# /// The data is downloaded from the web from the [CVDF mirror](https://github.com/cvdfoundation/mnist).
```

Comment:
you have leading "#"

Reply:
The leading `#` is the mdbook syntax for code that can be hidden. It can be toggled in the code block when the book is rendered.

Didn't want to have too much code overload the reader.